File: flutter/pid_reader.dart
import 'pid_type.dart';

/// PID Reader interface - OBD-II bağlantısı için
abstract class PIDReader {
  /// PID oku ve parse edilmiş değer döndür
  Future<double> readPID(PIDType pidType);

  /// Ham OBD komutu gönder
  Future<String> sendRawCommand(String command, String header);

  /// Bağlantı durumu
  Future<bool> isConnected();

  /// VIN oku
  Future<String> readVIN() async {
    try {
      return await sendRawCommand('09 02', '7DF');
    } catch (e) {
      return '';
    }
  }

  /// ECU adını oku
  Future<String> readECUName() async {
    try {
      return await sendRawCommand('09 0A', '7DF');
    } catch (e) {
      return '';
    }
  }
}

/// Mock PID Reader - Test için
class MockPIDReader implements PIDReader {
  final Map<PIDType, double> _mockData = {};

  MockPIDReader() {
    // Varsayılan değerler
    _mockData[PIDType.engineRpm] = 2000.0;
    _mockData[PIDType.vehicleSpeed] = 60.0;
    _mockData[PIDType.throttlePosition] = 45.0;
    _mockData[PIDType.engineLoad] = 35.0;
    _mockData[PIDType.coolantTemp] = 85.0;
    _mockData[PIDType.intakeTemp] = 25.0;
    _mockData[PIDType.fuelRate] = 8.5; // ✅ Gerçek yakıt tüketimi
    _mockData[PIDType.maf] = 15.0;
    _mockData[PIDType.batterySoc] = 75.0;
    _mockData[PIDType.batteryVoltage] = 360.0;
    _mockData[PIDType.batteryCurrent] = -20.0;
    _mockData[PIDType.batteryTemp] = 30.0;
    _mockData[PIDType.regenPower] = 15.0; // ✅ Regen gücü
    _mockData[PIDType.regenCurrent] = 40.0;
    _mockData[PIDType.motorTorque] = -50.0;
    _mockData[PIDType.regenEnergyTotal] = 125.5;
  }

  @override
  Future<double> readPID(PIDType pidType) async {
    await Future.delayed(const Duration(milliseconds: 50)); // Simüle gecikme
    return _mockData[pidType] ?? 0.0;
  }

  @override
  Future<String> sendRawCommand(String command, String header) async {
    await Future.delayed(const Duration(milliseconds: 50));

    // Mock responses
    if (command.contains('09 02')) {
      return '1HGCM82633A123456'; // Honda VIN
    } else if (command.contains('09 0A')) {
      return 'HONDA-ECU-2023';
    } else if (command.contains('22 015B')) {
      return '62 015B 64'; // Nissan yanıtı
    }

    return 'NO DATA';
  }

  @override
  Future<bool> isConnected() async {
    return true;
  }

  /// Test için mock değer ayarla
  void setMockValue(PIDType pidType, double value) {
    _mockData[pidType] = value;
  }
}

File: PIDReader.kt
package com.obd.diagnostics.pid

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * PIDReader interface - implement this for your OBD-II connection
 * (Bluetooth, USB, WiFi, etc.)
 */
interface PIDReader {
    /**
     * Read a PID and return parsed value
     */
    suspend fun readPID(pidType: PIDType): Double
    
    /**
     * Send raw OBD command and get response
     */
    suspend fun sendRawCommand(command: String, header: String = "7DF"): String
    
    /**
     * Check if connection is active
     */
    suspend fun isConnected(): Boolean
}

/**
 * Mock implementation for testing
 */
class MockPIDReader : PIDReader {
    private var mockData = mutableMapOf<PIDType, Double>()
    
    override suspend fun readPID(pidType: PIDType): Double = withContext(Dispatchers.IO) {
        mockData.getOrPut(pidType) {
            when (pidType) {
                PIDType.ENGINE_RPM -> 2000.0
                PIDType.VEHICLE_SPEED -> 60.0
                PIDType.THROTTLE_POSITION -> 45.0
                PIDType.ENGINE_LOAD -> 35.0
                PIDType.COOLANT_TEMP -> 85.0
                PIDType.INTAKE_TEMP -> 25.0
                PIDType.FUEL_RATE -> 8.5 // Real fuel consumption
                PIDType.MAF -> 15.0
                PIDType.BATTERY_SOC -> 75.0
                PIDType.BATTERY_VOLTAGE -> 360.0
                PIDType.BATTERY_CURRENT -> -20.0
                PIDType.BATTERY_TEMP -> 30.0
                PIDType.REGEN_POWER -> 15.0 // Regenerating 15 kW
                PIDType.REGEN_CURRENT -> 40.0
                PIDType.MOTOR_TORQUE -> -50.0
                else -> 0.0
            }
        }
    }
    
    override suspend fun sendRawCommand(command: String, header: String): String = 
        withContext(Dispatchers.IO) {
            // Mock responses for VIN and ECU
            when {
                command.contains("09 02") -> "1HGCM82633A123456" // Mock VIN (Honda)
                command.contains("09 0A") -> "HONDA-ECU-2023"
                command.contains("22 015B") -> "62 015B 64" // Nissan response
                else -> "NO DATA"
            }
        }
    
    override suspend fun isConnected(): Boolean = true
    
    fun setMockValue(pidType: PIDType, value: Double) {
        mockData[pidType] = value
    }
}
